Implement _WKSerializedNode for more Node types

achristensen07 opened the pull request:
#### 90a2c4a82bf83409efba2236f83be47eca99fa51
<pre>
Implement _WKSerializedNode for more Node types
<a href="https://bugs.webkit.org/show_bug.cgi?id=296299">https://bugs.webkit.org/show_bug.cgi?id=296299</a>
<a href="https://rdar.apple.com/156350776">rdar://156350776</a>

Reviewed by Ryosuke Niwa.

This expands from just Text nodes to all the nodes that
are easy to implement serialization of.

* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::serializeNode const):
* Source/WebCore/dom/CDATASection.cpp:
(WebCore::CDATASection::serializeNode const):
* Source/WebCore/dom/Comment.cpp:
(WebCore::Comment::serializeNode const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::serializeNode const):
* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::serializeNode const):
* Source/WebCore/dom/DocumentType.cpp:
(WebCore::DocumentType::serializeNode const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::serializeNode const):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::deserializeNode):
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::serializeNode const):
* Source/WebCore/dom/SerializedNode.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::serializeNode const):
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::serializeNode const):
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::serializeNode const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::TEST(SerializedNode, Basic)):

Canonical link: <a href="https://commits.webkit.org/297722@main">https://commits.webkit.org/297722@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e49a2cb14bb91cdcdff2e70d44edb23baa70df18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32382 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22860 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118849 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63124 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114612 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40945 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85727 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36352 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115597 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26356 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101333 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66035 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25651 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19469 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62608 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95743 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19544 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122070 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39724 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29595 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94592 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40107 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97570 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94334 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39451 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17259 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35812 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18141 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39612 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45100 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39251 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42584 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40990 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Attr.cpp:
```diff
@@ -119,6 +119,16 @@ Ref<Node> Attr::cloneNodeInternal(Document& document, CloningOperation, CustomEl
 
+SerializedNode Attr::serializeNode(CloningOperation) const
+{
+    return { SerializedNode::Attr {
```
Can we put `SerializedNode::Attr` on a new line?

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -830,8 +832,23 @@ ExceptionOr<void> Node::normalize()
+    }, [&] (SerializedNode::DocumentType&& type) -> RefPtr<Node> {
+        return DocumentType::create(document, type.name, type.publicId, type.systemId);
+    }, [&] (SerializedNode::Comment&& comment) -> RefPtr<Node> {
+        return Comment::create(document, WTFMove(comment.data));
```
It seems a bit wasteful to have different variants for text, comment, cdata section, etc...
How about we just introduce an enum class like `CharacterDataNodeType`?

achristensen07 replied:
Comment doesn't inherit from Text.  I could express the inheritance relationship of CDATASection and Text though.

rniwa replied:
They all inherit from `CharacterData`.

achristensen07 committed:
Implement _WKSerializedNode for more Node types
https://bugs.webkit.org/show_bug.cgi?id=296299
rdar://156350776

Reviewed by Ryosuke Niwa.

This expands from just Text nodes to all the nodes that
are easy to implement serialization of.

* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::serializeNode const):
* Source/WebCore/dom/CDATASection.cpp:
(WebCore::CDATASection::serializeNode const):
* Source/WebCore/dom/Comment.cpp:
(WebCore::Comment::serializeNode const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::serializeNode const):
* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::serializeNode const):
* Source/WebCore/dom/DocumentType.cpp:
(WebCore::DocumentType::serializeNode const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::serializeNode const):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::deserializeNode):
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::serializeNode const):
* Source/WebCore/dom/SerializedNode.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::serializeNode const):
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::serializeNode const):
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::serializeNode const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::TEST(SerializedNode, Basic)):

Canonical link: https://commits.webkit.org/297722@main
